transformer_engine/pytorch/module/linear.py

```diff
@@ -397,7 +397,11 @@ def forward(
             ctx.grad_output_quantizer = grad_output_quantizer
             ctx.fuse_wgrad_accumulation = fuse_wgrad_accumulation
             if fuse_wgrad_accumulation and weight.requires_grad:
-                ctx.main_grad = weight.main_grad
+                # This check is needed to ensure that main_grad is not created
+                # during the forward pass when using MCore FSDP as it creates
+                # the main_grad buffer lazily before backprop
+                if not hasattr(param, "__fsdp_param__"):
+                    ctx.main_grad = weight.main_grad
```

Comment:
Instead of spreading out the conditional logic in multiple places, how about something like:
```suggestion
                if hasattr(weight, "__fsdp_param__"):
                    # MCore FSDP creates main_grad lazily before backward
                    ctx.main_grad_func = weight.get_main_grad
                else:
                    ctx.main_grad_func = lambda: weight.main_grad
```
Then we can just call the function directly in backward:
```python
main_grad = None
if ctx.fuse_wgrad_accumulation and ctx.requires_wgrad:
    main_grad = ctx.main_grad_func()
```